{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "unlikely-thomson",
   "metadata": {},
   "source": [
    "### Développement d’un pipeline de calcul du TMB\n",
    "\n",
    "#### Projet 4BiM, 2021"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "premier-shepherd",
   "metadata": {},
   "source": [
    "#### Auteurs : Marie Casimir, Loup Petitjean et Nicolas Mendiboure\n",
    "#### Encadrantes Innate-Pharma : Sabrina Carpentier et Luciana Bastista\n",
    "#### Encadrante INSA : Maïwenn Pineau"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adopted-description",
   "metadata": {},
   "source": [
    "*Pour install cyvcf2 : pip install cyvcf2*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "proper-night",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cyvcf2\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "oriented-montreal",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AN</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>BQ</th>\n",
       "      <th>...</th>\n",
       "      <th>MT</th>\n",
       "      <th>NS</th>\n",
       "      <th>DP</th>\n",
       "      <th>VT</th>\n",
       "      <th>SS</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>INDEL</th>\n",
       "      <th>START</th>\n",
       "      <th>STOP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29651735</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>0.1</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>3</td>\n",
       "      <td>413</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA,RNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>144852566</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.08</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>3</td>\n",
       "      <td>51</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>144951880</td>\n",
       "      <td>A</td>\n",
       "      <td>[T]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.15</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>148342496</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>82</td>\n",
       "      <td>0.07</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>3</td>\n",
       "      <td>1187</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>149340855</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.1</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>2</td>\n",
       "      <td>42</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>155294949</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, C, T]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>4</td>\n",
       "      <td>(37, 1, 1)</td>\n",
       "      <td>(0.09000000357627869, 0.0, 0.0)</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>3</td>\n",
       "      <td>409</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA,RNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>205038731</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.28</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>91888009</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0.1</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>2</td>\n",
       "      <td>48</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>100916306</td>\n",
       "      <td>A</td>\n",
       "      <td>[G, C]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>(125, 10)</td>\n",
       "      <td>(0.9300000071525574, 0.07000000029802322)</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>2</td>\n",
       "      <td>135</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>111118911</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.06</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>SOM</td>\n",
       "      <td>3</td>\n",
       "      <td>96</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM        POS REF        ALT  QUAL FILTER  AN          AC  \\\n",
       "0     1   29651735   C        [T]   0.0   None   2          40   \n",
       "1     1  144852566   G        [A]   0.0   None   2           4   \n",
       "2     1  144951880   A        [T]   0.0   None   2           6   \n",
       "3     1  148342496   C        [G]   0.0   None   2          82   \n",
       "4     1  149340855   T        [C]   0.0   None   2           4   \n",
       "5     1  155294949   G  [A, C, T]   0.0   None   4  (37, 1, 1)   \n",
       "6     1  205038731   C        [T]   0.0   None   2           8   \n",
       "7     2   91888009   C        [T]   0.0   None   2           5   \n",
       "8     2  100916306   A     [G, C]   0.0   None   3   (125, 10)   \n",
       "9     2  111118911   C        [T]   0.0   None   2           6   \n",
       "\n",
       "                                          AF  BQ  ...   MT  NS    DP   VT  SS  \\\n",
       "0                                        0.1  36  ...  SOM   3   413  SNP   2   \n",
       "1                                       0.08  37  ...  SOM   3    51  SNP   2   \n",
       "2                                       0.15  33  ...  SOM   2    40  SNP   2   \n",
       "3                                       0.07  37  ...  SOM   3  1187  SNP   2   \n",
       "4                                        0.1  32  ...  SOM   2    42  SNP   2   \n",
       "5            (0.09000000357627869, 0.0, 0.0)  33  ...  SOM   3   409  SNP   2   \n",
       "6                                       0.28  35  ...  SOM   2    29  SNP   2   \n",
       "7                                        0.1  37  ...  SOM   2    48  SNP   2   \n",
       "8  (0.9300000071525574, 0.07000000029802322)  34  ...  SOM   2   135  SNP   2   \n",
       "9                                       0.06  36  ...  SOM   3    96  SNP   2   \n",
       "\n",
       "    ORIGIN SOMATIC  INDEL START  STOP  \n",
       "0  DNA,RNA    True      0     5     3  \n",
       "1      DNA    True      0     0     0  \n",
       "2      DNA    True      0     2     1  \n",
       "3      DNA    True      0     7     7  \n",
       "4      DNA    True      0     0     2  \n",
       "5  DNA,RNA    True      0     2    11  \n",
       "6      DNA    True      0     0     0  \n",
       "7      DNA    True      0     1     0  \n",
       "8      DNA    True      0     2     3  \n",
       "9      DNA    True      0     0     2  \n",
       "\n",
       "[10 rows x 23 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Importation d'un VCF avec la librairie cyvcf2 :\n",
    "\n",
    "#Général :\n",
    "\n",
    "CHROM = []\n",
    "POS = []\n",
    "REF = []\n",
    "ALT = []\n",
    "QUAL = []\n",
    "FILTER = []\n",
    "\n",
    "# Détails de la section INFO du VCF :\n",
    "\n",
    "AN = [] #Total number of alleles in called genotypes\n",
    "AC = [] #Allele count in genotypes, for each ALT allele, in the same order as listed\n",
    "AF = [] #Allele Frequency in primary data, for each ALT allele, in the same order as listed\n",
    "BQ = [] #RMS base quality\n",
    "SB = [] #Strand bias\n",
    "FA = [] #Overall fraction of reads supporting ALT\n",
    "MC = [] #Modification base changes at this position\n",
    "MT = [] #Modification types at this position\n",
    "NS = [] #Number of Samples With Data\n",
    "DP = [] #Total Depth across samples\n",
    "VT = [] #Variant type, can be SNP, INS or DEL\n",
    "SS = [] #Variant status relative to non-adjacent Normal,0=wildtype,1=germline,2=somatic,3=LOH,4=post-transcriptional modification,5=unknown\n",
    "ORIGIN = [] #Where the call originated from, the tumor DNA, RNA, or both\n",
    "SOMATIC = [] #Indicates if record is a somatic mutation\n",
    "INDEL = [] #Number of indels for all samples\n",
    "START = [] #Number of reads starting at this position across all samples\n",
    "STOP = [] #Number of reads stopping at this position across all samples\n",
    "\n",
    "\n",
    "\n",
    "for record in cyvcf2.VCF(\"./test_radia.vcf\"):\n",
    "    CHROM.append(record.CHROM)\n",
    "    POS.append(record.POS)\n",
    "    REF.append(record.REF)\n",
    "    ALT.append(record.ALT)\n",
    "    QUAL.append(record.QUAL)\n",
    "    FILTER.append(record.FILTER)\n",
    "    \n",
    "    AN.append(record.INFO.get(\"AN\"))\n",
    "    AC.append(record.INFO.get(\"AC\"))\n",
    "    AF.append(record.INFO.get(\"AF\"))\n",
    "    BQ.append(record.INFO.get(\"BQ\"))\n",
    "    SB.append(record.INFO.get(\"SB\"))\n",
    "    FA.append(record.INFO.get(\"FA\"))\n",
    "    MC.append(record.INFO.get(\"MC\"))\n",
    "    MT.append(record.INFO.get(\"MT\"))\n",
    "    NS.append(record.INFO.get(\"NS\"))\n",
    "    DP.append(record.INFO.get(\"DP\"))\n",
    "    VT.append(record.INFO.get(\"VT\"))\n",
    "    SS.append(record.INFO.get(\"SS\"))\n",
    "    ORIGIN.append(record.INFO.get(\"ORIGIN\"))\n",
    "    SOMATIC.append(record.INFO.get(\"SOMATIC\"))\n",
    "    INDEL.append(record.INFO.get(\"INDEL\"))\n",
    "    START.append(record.INFO.get(\"START\"))\n",
    "    STOP.append(record.INFO.get(\"STOP\"))\n",
    "    \n",
    "df = pd.DataFrame(list(zip(CHROM, POS, REF, ALT, QUAL, FILTER)), \n",
    "                  columns=[\"CHROM\", \"POS\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\"])\n",
    "\n",
    "df[ [\"AN\", \"AC\", \"AF\", \"BQ\", \"SB\", \"FA\", \"MC\", \n",
    "     \"MT\", \"NS\", \"DP\", \"VT\", \"SS\", \"ORIGIN\", \n",
    "     \"SOMATIC\", \"INDEL\", \"START\", \"STOP\"]] = pd.DataFrame(list(zip (AN, AC, AF, BQ, SB, FA, MC,\n",
    "                                                                    MT, NS, DP, VT, SS, ORIGIN, \n",
    "                                                                    SOMATIC, INDEL, START, STOP)))\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lucky-mentor",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "structural-sigma",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "scenic-attitude",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unusual-cleanup",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "creative-sleeve",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "outer-wrestling",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
